diff -Naurp lshell-0.9.16.python3-orig/lshell/checkconfig.py lshell-0.9.16.python3/lshell/checkconfig.py
--- lshell-0.9.16.python3-orig/lshell/checkconfig.py	2020-08-10 15:57:12.297242732 -0500
+++ lshell-0.9.16.python3/lshell/checkconfig.py	2020-08-10 15:58:57.860307296 -0500
@@ -358,6 +358,8 @@ class CheckConfig:
         # get 'default' configuration if any
         self.get_config_sub('default')
 
+        # PE-3987 work around to db access issues, values grabbed from util/libnss-pxusers/src/common.h
+        rolelist = { 2106: "Auditor", 2107: "Admin", 2108: "Migrator", 2109: "Operator", 2110: "Monitor" }
         # get groups configuration if any.
         # for each group the user belongs to, check if specific configuration
         # exists.  The primary group has the highest priority.
@@ -369,6 +371,10 @@ class CheckConfig:
                 section = 'grp:' + grpname
                 self.get_config_sub(section)
             except KeyError:
+                if gid in rolelist:
+                    section = 'grp:' + rolelist[gid]
+                    self.get_config_sub(section)
+
                 pass
 
         # get user configuration if any
