--- a/drivers/md/md.c	2007-03-27 15:07:20.000000000 -0500
+++ obj-3d3000/drivers/md/md.c	2007-04-24 11:25:36.339620344 -0500
@@ -2107,7 +2107,7 @@ static void analyze_sbs(mddev_t * mddev)
 
 	if (mddev->recovery_cp != MaxSector &&
 	    mddev->level >= 1)
-		printk(KERN_ERR "md: %s: raid array is not clean"
+		printk(KERN_WARNING "md: %s: raid array is not clean"
 		       " -- starting background reconstruction\n",
 		       mdname(mddev));
 
