diff -ru old/include/scsi/scsi.h new/include/scsi/scsi.h
--- old/include/scsi/scsi.h	2006-03-19 23:53:29.000000000 -0600
+++ new/include/scsi/scsi.h	2006-05-23 13:56:45.801385592 -0500
@@ -14,7 +14,7 @@
  *	The maximum sg list length SCSI can cope with
  *	(currently must be a power of 2 between 32 and 256)
  */
-#define SCSI_MAX_PHYS_SEGMENTS	MAX_PHYS_SEGMENTS
+#define SCSI_MAX_PHYS_SEGMENTS	256
 
 
 /*
