diff -ru old/include/scsi/sg.h new/include/scsi/sg.h
--- old/include/scsi/sg.h	2006-03-19 23:53:29.000000000 -0600
+++ new/include/scsi/sg.h	2006-05-23 14:18:24.495820267 -0500
@@ -124,6 +124,7 @@
 #define SG_FLAG_UNUSED_LUN_INHIBIT 2   /* default is overwrite lun in SCSI */
 				/* command block (when <= SCSI_2) */
 #define SG_FLAG_MMAP_IO 4       /* request memory mapped IO */
+#define SG_FLAG_DIRECT_ONLY 8   /* error the request if direct I/O cannot be used */
 #define SG_FLAG_NO_DXFER 0x10000 /* no transfer of kernel buffers to/from */
 				/* user space (debug indirect IO) */
 
