diff --git a/lhr/meta.cpp b/lhr/meta.cpp
index 350b202..ae2e38b 100644
--- a/lhr/meta.cpp
+++ b/lhr/meta.cpp
@@ -138,7 +138,12 @@ int xp_metainfo::setfromsmbattr(errbuf &eb, int fd, std::string &m1, std::string
        }
 
        smbattrib = xp_metainfo::getAttrBits( _b );
-       LOG("attr bits - raw %#x, selected %#x", _b, smbattrib);
+       {
+           char device_path[PATH_MAX];
+           snprintf(device_path, PATH_MAX, "/proc/self/fd/%d", fd);
+           readlink(device_path, device_path, PATH_MAX);
+           LOG("info attr bits %s - raw %#x, selected %#x", device_path, _b, smbattrib);
+       }
        cflags |= XPMI_SMBATTRIB;
        return 0;
