--- a/include/linux/blkdev.h	2006-03-19 23:53:29.000000000 -0600
+++ obj-750/include/linux/blkdev.h	2006-05-24 13:30:56.299554105 -0500
@@ -793,8 +793,8 @@ extern long blk_congestion_wait(int rw, 
 extern void blk_rq_bio_prep(request_queue_t *, struct request *, struct bio *);
 extern int blkdev_issue_flush(struct block_device *, sector_t *);
 
-#define MAX_PHYS_SEGMENTS 128
-#define MAX_HW_SEGMENTS 128
+#define MAX_PHYS_SEGMENTS 1024	//128
+#define MAX_HW_SEGMENTS 1024	//128
 #define SAFE_MAX_SECTORS 255
 #define BLK_DEF_MAX_SECTORS 1024
 
--- a/include/scsi/scsi_host.h	2006-03-19 23:53:29.000000000 -0600
+++ obj-750/include/scsi/scsi_host.h	2006-05-24 13:24:23.640165066 -0500
@@ -30,7 +30,7 @@ struct scsi_transport_template;
  *	 used in one scatter-gather request.
  */
 #define SG_NONE 0
-#define SG_ALL 0xff
+#define SG_ALL 0x3fff
 
 
 #define DISABLE_CLUSTERING 0
--- a/drivers/scsi/scsi_lib.c	2006-03-19 23:53:29.000000000 -0600
+++ obj-750/drivers/scsi/scsi_lib.c	2006-06-21 10:04:07.087469282 -0500
@@ -56,11 +56,17 @@ static struct scsi_host_sg_pool scsi_sg_
 #if (SCSI_MAX_PHYS_SEGMENTS > 128)
 	SP(256),
 #if (SCSI_MAX_PHYS_SEGMENTS > 256)
+	SP(512),
+#if (SCSI_MAX_PHYS_SEGMENTS > 512)
+	SP(1024),
+#if (SCSI_MAX_PHYS_SEGMENTS > 1024)
 #error SCSI_MAX_PHYS_SEGMENTS is too large
 #endif
 #endif
 #endif
 #endif
+#endif
+#endif
 }; 	
 #undef SP
 
@@ -750,6 +754,16 @@ static struct scatterlist *scsi_alloc_sg
 	case 129 ... 256:
 		cmd->sglist_len = 5;
 		break;
+#if (SCSI_MAX_PHYS_SEGMENTS  > 256)
+	case 257 ... 512:
+		cmd->sglist_len = 6;
+		break;
+#if (SCSI_MAX_PHYS_SEGMENTS  > 512)
+	case 513 ... 1024:
+		cmd->sglist_len = 7;
+		break;
+#endif
+#endif
 #endif
 #endif
 #endif
