--- a/include/linux/i2c-id.h	2007-02-04 12:44:54.000000000 -0600
+++ obj-3000/include/linux/i2c-id.h	2009-03-09 15:40:58.000000000 -0500
@@ -157,6 +157,7 @@
 #define I2C_DRIVERID_FSCHER 1046
 #define I2C_DRIVERID_W83L785TS 1047
 #define I2C_DRIVERID_OV7670 1048	/* Omnivision 7670 camera */
+#define I2C_DRIVERID_LM93 1049
 
 /*
  * ---- Adapter types ----------------------------------------------------
