--- a/arch/i386/kernel/mpparse.c	2006-11-29 15:57:37.000000000 -0600
+++ obj-3d3000/arch/i386/kernel/mpparse.c	2007-06-06 16:41:06.635814335 -0500
@@ -180,9 +180,6 @@ static void __devinit MP_processor_info 
 	}
 	apic_version[m->mpc_apicid] = ver;
 
-	phys_cpu = apicid_to_cpu_present(apicid);
-	physids_or(phys_cpu_present_map, phys_cpu_present_map, phys_cpu);
-
 	if (num_processors >= NR_CPUS) {
 		printk(KERN_WARNING "WARNING: NR_CPUS limit of %i reached."
 			"  Processor ignored.\n", NR_CPUS);
@@ -195,6 +192,9 @@ static void __devinit MP_processor_info 
 		return;
 	}
 
+	phys_cpu = apicid_to_cpu_present(apicid);
+	physids_or(phys_cpu_present_map, phys_cpu_present_map, phys_cpu);
+
 	cpu_set(num_processors, cpu_possible_map);
 	num_processors++;
 
