diff -ru old/include/scsi/sg.h new/include/scsi/sg.h
--- old/include/scsi/sg.h	2006-03-19 23:53:29.000000000 -0600
+++ new/include/scsi/sg.h	2006-05-23 14:12:18.572941621 -0500
@@ -236,7 +236,11 @@
 #define SG_DEF_RESERVED_SIZE SG_SCATTER_SZ /* load time option */
 
 /* maximum outstanding requests, write() yields EDOM if exceeded */
-#define SG_MAX_QUEUE 16
+/* since kmalloc uses power-of-two buckets, SG_MAX_QUEUE increases might as
+ * well bring the Sg_fd size near a power of 2.
+ * SG_MAX_QUEUE 84 => sizeof(Sg_fd) == 16200 on x86
+ */
+#define SG_MAX_QUEUE 84
 
 #define SG_BIG_BUFF SG_DEF_RESERVED_SIZE    /* for backward compatibility */
 
