--- elfutils-0.185/src/elflint.c	2021-10-18 16:08:29.665767794 -0500
+++ elfutils-0.185/src/elflint.c.new	2021-10-18 16:08:20.858792900 -0500
@@ -3430,15 +3430,16 @@ section [%2d] '%s': unknown parent versi
 
 static inline size_t
 buffer_pos (Elf_Data *data, const unsigned char *p)
 {
   return p - (const unsigned char *) data->d_buf;
 }
 
-inline size_t
+// inline size_t
+size_t
 buffer_left (Elf_Data *data, const unsigned char *p)
 {
   return (const unsigned char *) data->d_buf + data->d_size - p;
 }
 
 static void
 check_attributes (Ebl *ebl, GElf_Ehdr *ehdr, GElf_Shdr *shdr, int idx)
