diff -Nrp -U 7 --no-dereference --no-ignore-file-name-case --strip-trailing-cr '--exclude=.git*' rpm-4.17.0.orig/configure rpm-4.17.0/configure
--- rpm-4.17.0.orig/configure	2021-09-03 04:42:31.514342579 -0500
+++ rpm-4.17.0/configure	2021-10-14 15:32:51.107272947 -0500
@@ -25918,14 +25918,15 @@ else
   pkg_failed=yes
 fi
  else
     pkg_failed=untried
 fi
 if test -n "$LUA_LIBS"; then
     pkg_cv_LUA_LIBS="$LUA_LIBS"
+    pkg_failed=no
  elif test -n "$PKG_CONFIG"; then
     if test -n "$PKG_CONFIG" && \
     { { $as_echo "$as_me:${as_lineno-$LINENO}: \$PKG_CONFIG --exists --print-errors \"lua >= 5.3\""; } >&5
   ($PKG_CONFIG --exists --print-errors "lua >= 5.3") 2>&5
   ac_status=$?
   $as_echo "$as_me:${as_lineno-$LINENO}: \$? = $ac_status" >&5
   test $ac_status = 0; }; then
diff -Nrp -U 7 --no-dereference --no-ignore-file-name-case --strip-trailing-cr '--exclude=.git*' rpm-4.17.0.orig/docs/Makefile.am rpm-4.17.0/docs/Makefile.am
--- rpm-4.17.0.orig/docs/Makefile.am	2021-06-21 07:00:44.566611409 -0500
+++ rpm-4.17.0/docs/Makefile.am	2021-10-14 15:33:45.198118005 -0500
@@ -1,10 +1,10 @@
 ## Process this file with automake to produce Makefile.in
 
-SUBDIRS = man
+# SUBDIRS = man
 
 EXTRA_DIST =
 
 EXTRA_DIST += \
 _layouts/default.html \
 _layouts/favicon.ico \
 index.md \
diff -Nrp -U 7 --no-dereference --no-ignore-file-name-case --strip-trailing-cr '--exclude=.git*' rpm-4.17.0.orig/docs/Makefile.in rpm-4.17.0/docs/Makefile.in
--- rpm-4.17.0.orig/docs/Makefile.in	2021-09-03 04:42:31.655343581 -0500
+++ rpm-4.17.0/docs/Makefile.in	2021-10-14 15:33:54.353091781 -0500
@@ -428,15 +428,15 @@ sbindir = @sbindir@
 sharedstatedir = @sharedstatedir@
 srcdir = @srcdir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = man
+# SUBDIRS = man
 EXTRA_DIST = _layouts/default.html _layouts/favicon.ico index.md \
 	manual/index.md manual/arch_dependencies.md \
 	manual/autosetup.md manual/boolean_dependencies.md \
 	manual/buildprocess.md manual/file_triggers.md manual/lua.md \
 	manual/more_dependencies.md manual/scriptlet_expansion.md \
 	manual/buildroot.md manual/conditionalbuilds.md \
 	manual/dependencies.md manual/dependency_generators.md \
diff -Nrp -U 7 --no-dereference --no-ignore-file-name-case --strip-trailing-cr '--exclude=.git*' rpm-4.17.0.orig/configure rpm-4.17.0/configure
--- rpm-4.17.0.orig/configure	2021-09-03 04:42:31.514342579 -0500
+++ rpm-4.17.0/configure	2021-10-14 15:32:51.107272947 -0500
@@ -19218,15 +19218,15 @@ do
     $as_echo "$as_me:${as_lineno-$LINENO}: found $as_dir/$ac_word$ac_exec_ext" >&5
     break 2
   fi
 done
   done
 IFS=$as_save_IFS
 
+  test -z "$ac_cv_path___FIND_DEBUGINFO" && ac_cv_path___FIND_DEBUGINFO="${exec_prefix}/find-debuginfo"
-  test -z "$ac_cv_path___FIND_DEBUGINFO" && ac_cv_path___FIND_DEBUGINFO="/usr/bin/find-debuginfo"
   ;;
 esac
 fi
 __FIND_DEBUGINFO=$ac_cv_path___FIND_DEBUGINFO
 if test -n "$__FIND_DEBUGINFO"; then
   { $as_echo "$as_me:${as_lineno-$LINENO}: result: $__FIND_DEBUGINFO" >&5
 $as_echo "$__FIND_DEBUGINFO" >&6; }
