--- a/fs/hugetlbfs/inode.c	2008-05-02 16:20:34.451709735 -0500
+++ obj-3000/fs/hugetlbfs/inode.c	2008-05-05 12:12:45.000000000 -0500
@@ -92,7 +92,7 @@ static int hugetlbfs_file_mmap(struct fi
 	 * way when do_mmap_pgoff unwinds (may be important on powerpc
 	 * and ia64).
 	 */
-	vma->vm_flags |= VM_HUGETLB | VM_RESERVED;
+	vma->vm_flags |= VM_HUGETLB | VM_RESERVED | VM_ALWAYSDUMP;
 	vma->vm_ops = &hugetlb_vm_ops;
 
 	if (vma->vm_pgoff & ~(huge_page_mask(h) >> PAGE_SHIFT))
