--- a/fs/exec.c	2004-11-19 12:52:15.000000000 -0600
+++ b/fs/exec.c	2004-11-19 12:55:48.000000000 -0600
@@ -1507,7 +1507,7 @@
 	 * and core_uses_pid is set, then .%pid will be appended to
 	 * the filename. Do not do this for piped commands. */
 	if (!ispipe && !pid_in_pattern
-	    && (core_uses_pid || nr_threads)) {
+	    && (core_uses_pid)) {
 		rc = snprintf(out_ptr, out_end - out_ptr,
 			      ".%d", task_tgid_vnr(current));
 		if (rc > out_end - out_ptr)
