# $Id: scsi_proc.patch 15585 2006-11-30 22:59:47Z RustadM $

--- a/drivers/scsi/scsi_proc.c
+++ b/drivers/scsi/scsi_proc.c
@@ -347,8 +347,12 @@
 	 * Usage: echo "scsi add-single-device 0 1 2 3" >/proc/scsi/scsi
 	 * with  "0 1 2 3" replaced by your "Host Channel Id Lun".
 	 */
-	if (!strncmp("scsi add-single-device", buffer, 22)) {
-		p = buffer + 23;
+	p = buffer;
+	if (!strncmp("scsi ", p, 5) || !strncmp("scsi-", p, 5))
+		p += 5;	/* Skip leading "scsi " or "scsi-" if present */
+
+	if (!strncmp("add-single-device", p, 17)) {
+		p += 18;
 
 		host = simple_strtoul(p, &p, 0);
 		channel = simple_strtoul(p + 1, &p, 0);
@@ -361,8 +365,8 @@
 	 * Usage: echo "scsi remove-single-device 0 1 2 3" >/proc/scsi/scsi
 	 * with  "0 1 2 3" replaced by your "Host Channel Id Lun".
 	 */
-	} else if (!strncmp("scsi remove-single-device", buffer, 25)) {
-		p = buffer + 26;
+	} else if (!strncmp("remove-single-device", p, 20)) {
+		p += 21;
 
 		host = simple_strtoul(p, &p, 0);
 		channel = simple_strtoul(p + 1, &p, 0);
