--- a/drivers/watchdog/w83627hf_wdt.c	2008-10-09 17:13:53.000000000 -0500
+++ obj-7000/drivers/watchdog/w83627hf_wdt.c	2009-02-03 09:18:57.000000000 -0600
@@ -343,9 +343,11 @@ static int __init wdt_init(void)
 
 	ret = misc_register(&wdt_miscdev);
 	if (ret != 0) {
-		printk(KERN_ERR PFX
+		printk(KERN_WARNING PFX
 			"cannot register miscdev on minor=%d (err=%d)\n",
 							WATCHDOG_MINOR, ret);
+		printk(KERN_INFO PFX "Disabling watchdog\n");
+		wdt_disable();
 		goto unreg_reboot;
 	}
 
