diff -Nrp -U 7 --no-dereference --no-ignore-file-name-case --strip-trailing-cr '--exclude=.git*' gcc/binutils-gdb/gdb/configure.old gcc/binutils-gdb/gdb/configure
--- gcc/binutils-gdb/gdb/configure.old	2020-11-11 17:11:32.432427986 -0600
+++ gcc/binutils-gdb/gdb/configure	2020-11-11 17:12:00.099337964 -0600
@@ -11461,14 +11461,15 @@ if test "${enable_source_highlight+set}"
   yes)  enable_source_highlight=yes ;;
   no)   enable_source_highlight=no  ;;
   *)    as_fn_error $? "bad value ${enableval} for source-highlight option" "$LINENO" 5 ;;
 esac
 else
   enable_source_highlight=auto
 fi
+enable_source_highlight=no
 
 
 if test "${enable_source_highlight}" != "no"; then
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for the source-highlight library" >&5
 $as_echo_n "checking for the source-highlight library... " >&6; }
   if test "${pkg_config_prog_path}" = "missing"; then
     { $as_echo "$as_me:${as_lineno-$LINENO}: result: no - pkg-config not found" >&5
