--- include/error.h.original	2021-10-18 18:57:44.284393881 -0500
+++ include/error.h	2021-10-18 18:57:44.284393881 -0500
@@ -33,14 +33,14 @@
 
 /* If we know the function will never return make sure the compiler
    realizes that, too.  */
-__extern_always_inline void
-error (int __status, int __errnum, const char *__format, ...)
-{
-  if (__builtin_constant_p (__status) && __status != 0)
-    __error_noreturn (__status, __errnum, __format, __va_arg_pack ());
-  else
-    __error_alias (__status, __errnum, __format, __va_arg_pack ());
-}
+// __extern_always_inline void
+// error (int __status, int __errnum, const char *__format, ...)
+// {
+//   if (__builtin_constant_p (__status) && __status != 0)
+//     __error_noreturn (__status, __errnum, __format, __va_arg_pack ());
+//   else
+//     __error_alias (__status, __errnum, __format, __va_arg_pack ());
+// }
 
 
 extern void __REDIRECT (__error_at_line_alias, (int __status, int __errnum,
