--- a/drivers/scsi/aic94xx/aic94xx_sds.c	2006-06-22 17:35:03.648291710 -0500
+++ b/drivers/scsi/aic94xx/aic94xx_sds.c	2006-06-22 17:40:07.579207058 -0500
@@ -849,14 +849,14 @@ static int asd_process_ctrla_phy_setting
 		/* This is the SAS address which should be sent in IDENTIFY. */
 		memcpy(asd_ha->hw_prof.phy_desc[i].sas_addr, pe->sas_addr,
 		       SAS_ADDR_SIZE);
-		asd_ha->hw_prof.phy_desc[i].max_sas_lrate =
-			(pe->sas_link_rates & 0xF0) >> 4;
-		asd_ha->hw_prof.phy_desc[i].min_sas_lrate =
-			(pe->sas_link_rates & 0x0F);
-		asd_ha->hw_prof.phy_desc[i].max_sata_lrate =
-			(pe->sata_link_rates & 0xF0) >> 4;
-		asd_ha->hw_prof.phy_desc[i].min_sata_lrate =
-			(pe->sata_link_rates & 0x0F);
+		asd_ha->hw_prof.phy_desc[i].max_sas_lrate = PHY_LINKRATE_1_5;
+			//(pe->sas_link_rates & 0xF0) >> 4;
+		asd_ha->hw_prof.phy_desc[i].min_sas_lrate = PHY_LINKRATE_1_5;
+			//(pe->sas_link_rates & 0x0F);
+		asd_ha->hw_prof.phy_desc[i].max_sata_lrate = PHY_LINKRATE_1_5;
+			//(pe->sata_link_rates & 0xF0) >> 4;
+		asd_ha->hw_prof.phy_desc[i].min_sata_lrate = PHY_LINKRATE_1_5;
+			//(pe->sata_link_rates & 0x0F);
 		asd_ha->hw_prof.phy_desc[i].flags = pe->flags;
 		ASD_DPRINTK("ctrla: phy%d: sas_addr: %llx, sas rate:0x%x-0x%x,"
 			    " sata rate:0x%x-0x%x, flags:0x%x\n",
