--- a/drivers/scsi/sas/sas_scsi_host.c	2006-03-10 10:58:04.548781119 -0600
+++ b/drivers/scsi/sas/sas_scsi_host.c	2006-03-10 11:15:15.859944966 -0600
@@ -602,6 +602,10 @@
 		SAS_DPRINTK("sas device not found! How is this possible?\n");
 		return -ENODEV;
 	}
+	if (lu->inquiry_valid_data_len && (lu->inquiry_data[0] & 0x1f) == 0) {
+		scsi_dev->sdev_bflags = 0x100000;
+		printk(KERN_ERR "sas: No ULD for you!\n");
+	}
 	kobject_get(&lu->lu_obj);
 	return 0;
 }
