--- a/fs/binfmt_elf.c	2006-03-13 11:50:21.527264948 -0600
+++ obj-750/fs/binfmt_elf.c	2006-03-13 13:34:53.986649733 -0600
@@ -1154,9 +1154,11 @@ static int dump_seek(struct file *file, 
  */
 static int maydump(struct vm_area_struct *vma)
 {
-	/* Do not dump I/O mapped devices or special mappings */
-	if (vma->vm_flags & (VM_IO | VM_RESERVED))
+	/* Do not dump I/O mapped devices */
+	if (vma->vm_flags & VM_IO)
 		return 0;
+	if (vma->vm_flags & VM_RESERVED)
+		return 1;
 
 	/* Dump shared memory only if mapped from an anonymous file.  */
 	if (vma->vm_flags & VM_SHARED)
