--- a/fs/exec.c	2004-11-19 12:52:15.000000000 -0600
+++ b/fs/exec.c	2004-11-19 12:55:48.000000000 -0600
@@ -1358,7 +1358,7 @@
 	 * and core_uses_pid is set, then .%pid will be appended to
 	 * the filename */
 	if (!pid_in_pattern
-            && (core_uses_pid || atomic_read(&current->mm->mm_users) != 1)) {
+            && (core_uses_pid)) {
 		rc = snprintf(out_ptr, out_end - out_ptr,
 			      ".%d", current->tgid);
 		if (rc > out_end - out_ptr)
