--- a/drivers/hwmon/Makefile	2005-10-20 15:56:10.642608630 -0500
+++ b/drivers/hwmon/Makefile	2005-10-24 16:33:30.374012234 -0500
@@ -35,6 +35,8 @@
 obj-$(CONFIG_SENSORS_LM87)	+= lm87.o
 obj-$(CONFIG_SENSORS_LM90)	+= lm90.o
 obj-$(CONFIG_SENSORS_LM92)	+= lm92.o
+obj-$(CONFIG_SENSORS_LM93)	+= lm93.o
+obj-$(CONFIG_SENSORS_M1Z3)	+= m1z3.o
 obj-$(CONFIG_SENSORS_MAX1619)	+= max1619.o
 obj-$(CONFIG_SENSORS_PC87360)	+= pc87360.o
 obj-$(CONFIG_SENSORS_SIS5595)	+= sis5595.o
--- a/drivers/hwmon/Kconfig	2005-10-20 15:56:10.642608630 -0500
+++ b/drivers/hwmon/Kconfig	2005-10-24 16:33:07.770200880 -0500
@@ -291,6 +291,27 @@
 	  This driver can also be built as a module.  If so, the module
 	  will be called lm92.
 
+config SENSORS_LM93
+	tristate "National Semiconductor LM93"
+	depends on HWMON && I2C && EXPERIMENTAL
+	select HWMON_VID
+	help
+	  If you say yes here you get support for National Semiconductor LM93.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called lm93.
+
+config SENSORS_M1Z3
+	tristate "Zippy Redundant Switching Power Supply"
+	depends on HWMON && I2C && EXPERIMENTAL
+	select HWMON_VID
+	help
+	  If you say yes here you get support for Zippy Redundant Switching
+	  Power Supply.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called m1z3.
+
 config SENSORS_MAX1619
 	tristate "Maxim MAX1619 sensor chip"
 	depends on HWMON && I2C && EXPERIMENTAL
--- a/include/linux/i2c-id.h	2005-10-20 15:56:10.642608630 -0500
+++ b/include/linux/i2c-id.h	2005-10-24 16:33:30.374012234 -0500
@@ -151,6 +151,7 @@ #define I2C_DRIVERID_LM90 1042
 #define I2C_DRIVERID_ASB100 1043
 #define I2C_DRIVERID_FSCHER 1046
 #define I2C_DRIVERID_W83L785TS 1047
+#define I2C_DRIVERID_LM93 1049
 
 /*
  * ---- Adapter types ----------------------------------------------------

