diff -Nru old/drivers/scsi/iscsi_tcp.c new/drivers/scsi/iscsi_tcp.c
--- old/drivers/scsi/iscsi_tcp.c	2006-03-19 23:53:29.000000000 -0600
+++ new/drivers/scsi/iscsi_tcp.c	2006-05-23 10:52:10.047190407 -0500
@@ -2989,6 +2989,10 @@
 	struct iscsi_conn *conn = ctask->conn;
 	struct iscsi_session *session = conn->session;
 
+	/* Xiotech: disable eh_abort entirely */
+	printk("ignoring abort for [sc %lx itt 0x%x]\n", (long)sc, ctask->itt);
+	return SUCCESS;
+
 	conn->eh_abort_cnt++;
 	debug_scsi("aborting [sc %lx itt 0x%x]\n", (long)sc, ctask->itt);
 
