diff -rup old/drivers/scsi/scsi_transport_iscsi.c new/drivers/scsi/scsi_transport_iscsi.c
--- old/drivers/scsi/scsi_transport_iscsi.c	2006-03-19 23:53:29.000000000 -0600
+++ new/drivers/scsi/scsi_transport_iscsi.c	2006-06-22 15:45:48.743338703 -0500
@@ -31,8 +31,8 @@
 #include <scsi/scsi_transport_iscsi.h>
 #include <scsi/iscsi_if.h>
 
-#define ISCSI_SESSION_ATTRS 8
-#define ISCSI_CONN_ATTRS 6
+#define ISCSI_SESSION_ATTRS 9
+#define ISCSI_CONN_ATTRS 7
 
 struct iscsi_internal {
 	struct scsi_transport_template t;
@@ -1014,6 +1014,15 @@ iscsi_conn_int_attr(data_digest, ISCSI_P
 iscsi_conn_int_attr(ifmarker, ISCSI_PARAM_IFMARKER_EN, "%d");
 iscsi_conn_int_attr(ofmarker, ISCSI_PARAM_OFMARKER_EN, "%d");
 
+static ssize_t 
+show_conn_handle(struct class_device *cdev, char *buf)
+{
+	struct iscsi_cls_conn *conn = iscsi_cdev_to_conn(cdev);	\
+	return sprintf(buf, "%llu\n", (unsigned long long)iscsi_handle(conn));
+}
+static CLASS_DEVICE_ATTR(conn_handle, S_IRUGO, show_conn_handle, NULL);
+
+
 #define iscsi_cdev_to_session(_cdev) \
 	iscsi_dev_to_session(_cdev->dev)
 
@@ -1045,6 +1054,15 @@ iscsi_session_int_attr(data_pdu_in_order
 iscsi_session_int_attr(data_seq_in_order, ISCSI_PARAM_DATASEQ_INORDER_EN, "%d");
 iscsi_session_int_attr(erl, ISCSI_PARAM_ERL, "%d");
 
+static ssize_t 
+show_session_handle(struct class_device *cdev, char *buf)
+{
+	struct iscsi_cls_session *session = iscsi_cdev_to_session(cdev);	\
+	return sprintf(buf, "%llu\n", (unsigned long long)iscsi_handle(session));
+}
+static CLASS_DEVICE_ATTR(session_handle, S_IRUGO, show_session_handle, NULL);
+
+
 #define SETUP_SESSION_RD_ATTR(field, param)				\
 	if (priv->param_mask & (1 << param)) {				\
 		priv->session_attrs[count] = &class_device_attr_##field;\
@@ -1159,6 +1177,8 @@ iscsi_register_transport(struct iscsi_tr
 	SETUP_CONN_RD_ATTR(data_digest, ISCSI_PARAM_DATADGST_EN);
 	SETUP_CONN_RD_ATTR(ifmarker, ISCSI_PARAM_IFMARKER_EN);
 	SETUP_CONN_RD_ATTR(ofmarker, ISCSI_PARAM_OFMARKER_EN);
+	priv->conn_attrs[count] = &class_device_attr_conn_handle;
+	count++;
 
 	BUG_ON(count > ISCSI_CONN_ATTRS);
 	priv->conn_attrs[count] = NULL;
@@ -1178,6 +1198,8 @@ iscsi_register_transport(struct iscsi_tr
 	SETUP_SESSION_RD_ATTR(data_pdu_in_order, ISCSI_PARAM_PDU_INORDER_EN);
 	SETUP_SESSION_RD_ATTR(data_seq_in_order,ISCSI_PARAM_DATASEQ_INORDER_EN)
 	SETUP_SESSION_RD_ATTR(erl, ISCSI_PARAM_ERL);
+	priv->session_attrs[count] = &class_device_attr_session_handle;
+	count++;
 
 	BUG_ON(count > ISCSI_SESSION_ATTRS);
 	priv->session_attrs[count] = NULL;
