--- a/drivers/scsi/iscsi_tcp.c	2006-03-19 23:53:29.000000000 -0600
+++ obj-3d3000/drivers/scsi/iscsi_tcp.c	2006-06-02 12:52:35.404264121 -0500
@@ -2941,10 +2941,25 @@ static int
 iscsi_eh_host_reset(struct scsi_cmnd *sc)
 {
 	struct iscsi_cmd_task *ctask = (struct iscsi_cmd_task *)sc->SCp.ptr;
-	struct iscsi_conn *conn = ctask->conn;
-	struct iscsi_session *session = conn->session;
+	struct iscsi_session *session = NULL;
+	struct iscsi_conn *conn = NULL;
+
+	if (ctask) {
+		conn = ctask->conn;
+		session = conn->session;
+	}
+	else {
+		/* SG_SCSI_RESET ioctls pass commands with no associated
+		 * ctask, since the command never went through queuecommand.
+		 * FIXME: locking on hostdata/session/conn?
+		 */
+		session = iscsi_hostdata(sc->device->host->hostdata);
+		printk("iSCSI: ioctl requested host reset\n");
+	}
 
 	spin_lock_bh(&session->lock);
+	if (conn == NULL)
+		conn = session->leadconn;
 	if (session->state == ISCSI_STATE_TERMINATE) {
 		debug_scsi("failing host reset: session terminated "
 			   "[CID %d age %d]", conn->id, session->age);
