diff -ru a/drivers/scsi/libata-core.c override/drivers/scsi/libata-core.c
--- a/drivers/scsi/libata-core.c	2006-08-15 13:32:44.131589085 -0500
+++ b/drivers/scsi/libata-core.c	2006-08-10 17:19:39.000000000 -0500
@@ -1187,7 +1187,7 @@
 
 	spin_unlock_irqrestore(ap->lock, flags);
 
-	if (!wait_for_completion_timeout(&wait, ATA_TMOUT_INTERNAL)) {
+	if (!wait_for_completion_timeout(&wait, ATA_TMOUT_INTERNAL_QUICK)) {
 		spin_lock_irqsave(ap->lock, flags);
 
 		/* We're racing with irq here.  If we lose, the
