diff -rubB obj-750/drivers/scsi/libata-core.c obj-750.new/drivers/scsi/libata-core.c
--- obj-750/drivers/scsi/libata-core.c	2006-08-25 13:14:40.156001934 -0500
+++ obj-750.new/drivers/scsi/libata-core.c	2006-08-25 13:15:45.000000000 -0500
@@ -82,7 +82,7 @@
 module_param(atapi_enabled, int, 0444);
 MODULE_PARM_DESC(atapi_enabled, "Enable discovery of ATAPI devices (0=off, 1=on)");
 
-int ncq_enabled = 0;
+int ncq_enabled = 1;
 module_param(ncq_enabled, int, 0444);
 MODULE_PARM_DESC(ncq_enabled, "Enable NCQ (0=off, 1=on)");
 
diff -rubB obj-750/drivers/scsi/sas/sas_scsi_host.c obj-750.new/drivers/scsi/sas/sas_scsi_host.c
--- obj-750/drivers/scsi/sas/sas_scsi_host.c	2006-08-25 13:14:40.174000221 -0500
+++ obj-750.new/drivers/scsi/sas/sas_scsi_host.c	2006-08-25 13:09:11.000000000 -0500
@@ -43,6 +43,11 @@
 module_param_named(no_uld, sas_no_uld, int, S_IRUGO);
 MODULE_PARM_DESC(no_uld, "\n"
 		"\tPrevent upper level driver attach for disks, default:1\n");
+int    sas_qdepth = 1;
+module_param_named(qdepth, sas_qdepth, int, S_IRUGO);
+MODULE_PARM_DESC(qdepth, "\n"
+               "\tOverride LLDD queue depth. Default: 0 [no override]\n");
+
 
 /* The SAM LUN structure should be _completely_ opaque to SCSI Core.
  * This is why this macro here, and not using the broken
@@ -1136,7 +1141,7 @@
 	sht->module = sas_ha->lldd_module;
 	sht->name = sas_ha->sas_ha_name;
 	sht->can_queue = sas_ha->lldd_queue_size;
-	sht->cmd_per_lun = sas_ha->lldd_queue_size;
+        sht->cmd_per_lun = sas_qdepth == 0 ? sas_ha->lldd_queue_size : sas_qdepth;
 
 	sht->sg_tablesize = sas_ha->hw_profile.max_sg_table;
 	sht->max_sectors  = sas_ha->hw_profile.max_sectors;
