--- a/drivers/i2c/chips/Kconfig	2006-01-02 21:21:10.000000000 -0600
+++ b/drivers/i2c/chips/Kconfig	2006-02-20 13:10:52.715566661 -0600
@@ -46,6 +46,16 @@
 	  This driver can also be built as a module.  If so, the module
 	  will be called pcf8574.
 
+config SENSORS_PCA9551
+	tristate "Philips PCA9551 LED Blinker"
+	depends on I2C && EXPERIMENTAL
+	help
+	  If you say yes here you get support for the Philips PCA9551
+	  LED Blinker.
+
+	  This driver can also be built as a module.  If so, the module
+	  will be called pca9551.
+
 config SENSORS_PCA9539
 	tristate "Philips PCA9539 16-bit I/O port"
 	depends on I2C && EXPERIMENTAL
--- a/drivers/i2c/chips/Makefile	2006-01-02 21:21:10.000000000 -0600
+++ b/drivers/i2c/chips/Makefile	2006-02-20 13:13:15.873831747 -0600
@@ -7,6 +7,7 @@
 obj-$(CONFIG_SENSORS_EEPROM)	+= eeprom.o
 obj-$(CONFIG_SENSORS_MAX6875)	+= max6875.o
 obj-$(CONFIG_SENSORS_M41T00)	+= m41t00.o
+obj-$(CONFIG_SENSORS_PCA9551)	+= pca9551.o
 obj-$(CONFIG_SENSORS_PCA9539)	+= pca9539.o
 obj-$(CONFIG_SENSORS_PCF8574)	+= pcf8574.o
 obj-$(CONFIG_SENSORS_PCF8591)	+= pcf8591.o
