--- a/drivers/scsi/scsi_error.c	2006-03-19 23:53:29.000000000 -0600
+++ obj-3d3000/drivers/scsi/scsi_error.c	2006-08-30 13:56:42.494054624 -0500
@@ -1657,6 +1657,9 @@ scsi_reset_provider(struct scsi_device *
 	struct request req;
 	int rtn;
 
+	printk(KERN_INFO "scsi_reset_provider allocated cmd %p\n", scmd);
+	memset(&scmd->SCp, 0, sizeof(scmd->SCp));
+
 	scmd->request = &req;
 	memset(&scmd->eh_timeout, 0, sizeof(scmd->eh_timeout));
 	scmd->request->rq_status      	= RQ_SCSI_BUSY;
@@ -1702,7 +1705,8 @@ scsi_reset_provider(struct scsi_device *
 		rtn = FAILED;
 	}
 
-	scsi_next_command(scmd);
+	scsi_put_command(scmd);
+	printk(KERN_INFO "scsi_reset_provider cmd %p returning %d\n", scmd, rtn);
 	return rtn;
 }
 EXPORT_SYMBOL(scsi_reset_provider);
