diff -ru 2.6.16/obj-750/drivers/scsi/sas/sas_scsi_host.c /home/holtyb/cvsrep/head/work/Wookiee/Linux/xio3d/2.6.16/obj-750/drivers/scsi/sas/sas_scsi_host.c
--- a/drivers/scsi/sas/sas_scsi_host.c	2006-09-29 06:12:55.307093323 -0500
+++ b/drivers/scsi/sas/sas_scsi_host.c	2006-09-28 06:07:59.000000000 -0500
@@ -43,7 +43,7 @@
 module_param_named(no_uld, sas_no_uld, int, S_IRUGO);
 MODULE_PARM_DESC(no_uld, "\n"
 		"\tPrevent upper level driver attach for disks, default:1\n");
-int    sas_qdepth = 1;
+int    sas_qdepth = 0;
 module_param_named(qdepth, sas_qdepth, int, S_IRUGO);
 MODULE_PARM_DESC(qdepth, "\n"
                "\tOverride LLDD queue depth. Default: 0 [no override]\n");
@@ -816,6 +816,7 @@
 		break ;
 
 	case ATA_PROT_NCQ:
+        task->ata_task.fis.sector_count = (qc->tf.nsect & 0x7);
 		task->ata_task.use_ncq = 1;
 		/* Drop through */
 

