--- a/drivers/scsi/sas/sas_expander.c	2006-06-22 17:55:28.000000000 -0500
+++ b/drivers/scsi/sas/sas_expander.c	2006-06-23 10:24:46.747504443 -0500
@@ -605,6 +605,8 @@ static int smp_phy_control(struct domain
 	pc_req[1] = SMP_PHY_CONTROL;
 	pc_req[9] = phy_id;
 	pc_req[10]= phy_func;
+	if (phy_func == PHY_FUNC_LINK_RESET)
+		pc_req[33] = 0x80;
 
 	res = smp_execute_task(dev, pc_req, PC_REQ_SIZE, pc_resp,PC_RESP_SIZE);
 
