--- obj-750/drivers/scsi/sas/sas_discover.c	2006-09-22 15:41:54.255092856 -0500
+++ override/drivers/scsi/sas/sas_discover.c	2006-09-22 13:52:11.000000000 -0500
@@ -1575,9 +1575,11 @@
 				    __FUNCTION__, dev->dev_type);
 			break;
 		}
-                if (ret==-EAGAIN)
-                    SAS_DPRINTK("LOOK AT ME ret %02X ex_dev %p phy %p phyid %d\n",ret,ex_dev,phy,phy_id);
-                
+                if (ret==-EAGAIN) {
+                    sas_execute_disable_enable_phy(dev,0);
+                    SAS_DPRINTK("LOOK AT ME sas_register_xxx failed ret %02X ex_dev %p phy %p phyid %d\n"
+                            ,ret,ex_dev,phy,phy_id);
+                }
                 if ((parent) && (ret==-EAGAIN))
                 {
                     memset(phy->attached_sas_addr, 0, SAS_ADDR_SIZE);
