--- a/drivers/scsi/sas/sas_scsi_host.c	2006-03-23 08:42:19.204374362 -0600
+++ b/drivers/scsi/sas/sas_scsi_host.c	2006-03-23 08:41:01.967837526 -0600
@@ -863,7 +863,7 @@
 			lu->tm_type = (bq<<1) | cq;
 			scsi_dev->tagged_supported = 1;
 			if (cq)
-				scsi_set_tag_type(scsi_dev, MSG_ORDERED_TAG);
+				scsi_set_tag_type(scsi_dev, MSG_SIMPLE_TAG);
 			else
 				scsi_set_tag_type(scsi_dev, MSG_SIMPLE_TAG);
 			scsi_activate_tcq(scsi_dev, SAS_DEF_QD);
@@ -886,6 +886,8 @@
 		printk(KERN_ERR "sas: No ULD for you!\n");
 	}
 
+	blk_queue_dma_alignment(scsi_dev->request_queue, 0x0F);
+
 	if (dev->end_dev.itnl_timeout > 0)
 		scsi_dev->timeout = HZ +
 			msecs_to_jiffies(dev->end_dev.itnl_timeout);
