--- obj-750/drivers/scsi/sas/sas_scsi_host.c	2006-08-21 16:21:46.825900238 -0500
+++ override/drivers/scsi/sas/sas_scsi_host.c	2006-08-21 15:13:56.502636234 -0500
@@ -493,7 +493,13 @@
 
 	return res;
 }
+static int sas_dev_reset(struct scsi_cmnd *cmd)
+{
+    struct LU *lu = cmd->device->hostdata;
+    sas_execute_hard_reset(lu->parent);
 
+    return 0;
+}
 static int sas_scsi_recover_host(struct Scsi_Host *shost)
 {
 	struct sas_ha_struct *ha = SHOST_TO_SAS_HA(shost);
@@ -1108,6 +1113,7 @@
 	.queuecommand = sas_queuecommand,
 	.eh_strategy_handler = sas_scsi_recover_host,
 	.eh_timed_out = sas_scsi_timed_out,
+        .eh_device_reset_handler = sas_dev_reset,
 	.slave_alloc = sas_slave_alloc,
 	.slave_configure = sas_slave_configure,
 	.slave_destroy = sas_slave_destroy,
