--- a/drivers/scsi/sas/sas_expander.c	2006-06-20 15:54:33.259047428 -0500
+++ b/drivers/scsi/sas/sas_expander.c	2006-06-20 17:14:24.399976850 -0500
@@ -30,6 +30,8 @@
 #include <scsi/sas/sas_task.h>
 #include <scsi/sas/sas_discover.h>
 
+#include <linux/delay.h>	/*DWD*/
+
 static int sas_discover_expander(struct domain_device *dev);
 static int sas_configure_routing(struct domain_device *dev, u8 *sas_addr);
 static int sas_disable_routing(struct domain_device *dev,  u8 *sas_addr);
@@ -886,6 +888,8 @@ static int sas_ex_discover_dev(struct do
 
 	switch (ex_phy->attached_dev_type) {
 	case SAS_END_DEV:
+		res = smp_phy_control(dev, phy_id, PHY_FUNC_LINK_RESET);
+		mdelay(100) ;
 		child = sas_ex_discover_end_dev(dev, phy_id);
 		break;
 	case FANOUT_DEV:
