--- a/drivers/scsi/aic7xxx/aic79xx_osm.c	2006-02-27 09:35:27.273663294 -0600
+++ obj-3d3000/drivers/scsi/aic7xxx/aic79xx_osm.c	2006-02-28 15:49:19.948704805 -0600
@@ -719,6 +719,7 @@ struct scsi_host_template aic79xx_driver
 	.can_queue		= AHD_MAX_QUEUE,
 	.this_id		= -1,
 	.cmd_per_lun		= 2,
+	.max_sectors		= 0xffff,
 	.use_clustering		= ENABLE_CLUSTERING,
 	.slave_alloc		= ahd_linux_slave_alloc,
 	.slave_configure	= ahd_linux_slave_configure,
#--- a/drivers/scsi/aic94xx/aic94xx_init.c	2006-02-28 16:14:13.274737184 -0600
#+++ obj/drivers/scsi/aic94xx/aic94xx_init.c	2006-02-28 16:19:32.690343439 -0600
#@@ -483,7 +483,7 @@ static int asd_register_sas_ha(struct as
# 	asd_ha->sas_ha.sas_addr = &asd_ha->hw_prof.sas_addr[0];
# 
# 	asd_ha->sas_ha.hw_profile.max_sg_table = SG_ALL;
#-	asd_ha->sas_ha.hw_profile.max_sectors  = SCSI_DEFAULT_MAX_SECTORS;
#+	asd_ha->sas_ha.hw_profile.max_sectors  = 0xffff;
# 	asd_ha->sas_ha.hw_profile.dma_boundary = 0;
# 	asd_ha->sas_ha.hw_profile.use_clustering = ENABLE_CLUSTERING;
# 
