--- a/drivers/hwmon/w83627hf.c	2006-03-19 23:53:29.000000000 -0600
+++ obj-750/drivers/hwmon/w83627hf.c	2006-05-31 17:17:56.030041821 -0500
@@ -695,6 +695,27 @@ static DEVICE_ATTR(alarms, S_IRUGO, show
 #define device_create_file_alarms(client) \
 device_create_file(&client->dev, &dev_attr_alarms)
 
+static ssize_t
+show_powergood_reg(struct device *dev, struct device_attribute *attr,
+	char *buf)
+{
+	struct i2c_client *client = to_i2c_client(dev);
+	struct w83627hf_data *data = i2c_get_clientdata(client);
+	u8	crf1;
+
+	down(&data->update_lock);
+	superio_enter();
+	superio_select(W83627HF_LD_GPIO1);
+	crf1 = superio_inb(0xf1);
+	superio_exit();
+	up(&data->update_lock);
+
+	return sprintf(buf, "%d\n", (crf1 & 2) != 0);
+}
+static DEVICE_ATTR(powergood, S_IRUGO, show_powergood_reg, NULL);
+#define device_create_file_powergood(client) \
+device_create_file(&client->dev, &dev_attr_powergood)
+
 #define show_beep_reg(REG, reg) \
 static ssize_t show_beep_##reg (struct device *dev, struct device_attribute *attr, char *buf) \
 { \
@@ -1134,6 +1155,7 @@ static int w83627hf_detect(struct i2c_ad
 		device_create_file_fan_div(new_client, 3);
 
 	device_create_file_alarms(new_client);
+	device_create_file_powergood(new_client);
 
 	device_create_file_beep(new_client);
 
