--- obj-750/drivers/scsi/sas/sas_discover.c	2006-08-23 14:08:46.219361033 -0500
+++ override/drivers/scsi/sas/sas_discover.c	2006-08-23 13:17:21.614735135 -0500
@@ -1324,17 +1324,21 @@
     if (dev->dev_type == SATA_DEV) {
         scsi_dev = dev->sata_dev.lu->uldd_dev;
         if (scsi_dev && scsi_dev->request_queue) {
-            scsi_internal_device_block(scsi_dev);
+            
             if (cancel)
                 scsi_device_set_state(scsi_dev,SDEV_CANCEL);
+            else
+                scsi_internal_device_block(scsi_dev);
         }
     } else if (dev->dev_type == SAS_END_DEV){
         list_for_each_entry_safe(lu, n, &dev->end_dev.LU_list, list) {
             scsi_dev = lu->uldd_dev;
             if (scsi_dev && scsi_dev->request_queue) {
-                scsi_internal_device_block(scsi_dev);
+               
                 if (cancel)
                     scsi_device_set_state(scsi_dev,SDEV_CANCEL);
+                else
+                    scsi_internal_device_block(scsi_dev);
             }
         }
     } else {
