diff -ur ./rsyslog-6.4.2/configure.old ./sources/rsyslog-6.4.2/configure 
--- ./rsyslog-6.4.2/configure.old	2012-09-20 02:09:19.000000000 -0500
+++ ./rsyslog-6.4.2/configure	2012-09-20 02:09:19.000000000 -0500
@@ -13536,7 +13536,7 @@
 esac
 
 
-$as_echo "#define malloc rpl_malloc" >>confdefs.h
+# $as_echo "#define malloc rpl_malloc" >>confdefs.h
 
 fi
 
@@ -13603,7 +13603,7 @@
 esac
 
 
-$as_echo "#define realloc rpl_realloc" >>confdefs.h
+# $as_echo "#define realloc rpl_realloc" >>confdefs.h
 
 fi
 
