diff -ur ncurses.orig/include/curses.tail ncurses-6.0/include/curses.tail
--- ncurses.orig/include/curses.tail	2011-10-29 15:03:22.000000000 -0500
+++ ncurses-6.0/include/curses.tail	2017-01-08 13:41:27.836697482 -0600
@@ -111,7 +111,7 @@
 extern NCURSES_EXPORT(int)     NCURSES_SP_NAME(mouseinterval) (SCREEN*, int);
 #endif
 
-#define mouse_trafo(y,x,to_screen) wmouse_trafo(stdscr,y,x,to_screen)
+// #define mouse_trafo(y,x,to_screen) wmouse_trafo(stdscr,y,x,to_screen)
 
 /* other non-XSI functions */
 
