diff -ur ./procps-ng-3.3.11/configure ../sources/procps-ng-3.3.11/configure
--- ./procps-ng-3.3.11/configure	2015-08-09 02:57:06.537284846 -0500
+++ ../sources/procps-ng-3.3.11/configure	2017-01-05 09:35:48.894745117 -0600
@@ -17686,7 +17686,7 @@
 esac
 
 
-$as_echo "#define malloc rpl_malloc" >>confdefs.h
+# $as_echo "#define malloc rpl_malloc" >>confdefs.h
 
 fi
 
@@ -17775,7 +17775,7 @@
 /* end confdefs.h.  */
 $ac_includes_default
 /* malloc might have been renamed as rpl_malloc. */
-#undef malloc
+# #undef malloc
 
 /* Thanks to Mike Haertel and Jim Avera for this test.
    Here is a matrix of mmap possibilities:
@@ -17991,7 +17991,7 @@
 esac
 
 
-$as_echo "#define realloc rpl_realloc" >>confdefs.h
+# $as_echo "#define realloc rpl_realloc" >>confdefs.h
 
 fi
 
