diff -ur ./psmisc-22.21/configure ../sources/psmisc-22.21/configure
--- ./psmisc-22.21/configure	2014-01-31 06:53:58.000000000 -0600
+++ ../sources/psmisc-22.21/configure	2017-01-05 09:47:43.309978993 -0600
@@ -7487,7 +7487,7 @@
 esac
 
 
-$as_echo "#define malloc rpl_malloc" >>confdefs.h
+# $as_echo "#define malloc rpl_malloc" >>confdefs.h
 
 fi
 
@@ -7616,7 +7616,7 @@
 esac
 
 
-$as_echo "#define realloc rpl_realloc" >>confdefs.h
+# $as_echo "#define realloc rpl_realloc" >>confdefs.h
 
 fi
 
