diff -ur ./flex-2.5.39/configure ../sources/flex-2.5.39/configure
--- ./flex-2.5.39/configure	2014-03-26 13:58:55.000000000 -0500
+++ ../sources/flex-2.5.39/configure	2017-01-05 09:30:11.504531716 -0600
@@ -19318,7 +19318,7 @@
 esac
 
 
-$as_echo "#define malloc rpl_malloc" >>confdefs.h
+# $as_echo "#define malloc rpl_malloc" >>confdefs.h
 
 fi
 
@@ -19385,7 +19385,7 @@
 esac
 
 
-$as_echo "#define realloc rpl_realloc" >>confdefs.h
+# $as_echo "#define realloc rpl_realloc" >>confdefs.h
 
 fi
 
